### ISSUE TITLE
v1.5.0 verbs;ofi_rxm is only using loopback.

### ISSUE THREAD

**Issue by bcernohous** [ofiwg/libfabric#3261]
I was unsuccessfully trying to get an RMA program working on the latest libfabric using verbs (which works at v1.4.x).  The difference seems to be ofi_rxm.

I noticed the installed fi_pingpong was working so I tried using similar hints.  It seems FI_MSG (fi_pingpong) is working but FI_RMA is not working.

I use these basic hints (copied from pingpong.c)

        hints->ep_attr->type = FI_EP_RDM;
        hints->mode = FI_CONTEXT;
        hints->domain_attr->mr_mode = FI_MR_LOCAL |
            (FI_MR_ALLOCATED | FI_MR_PROV_KEY | FI_MR_VIRT_ADDR);

And then either 

        hints->caps = FI_MSG;

or

        hints->caps = FI_RMA;

For FI_MSG I get back both verbs and verbs;ofi_rxm with:

prov_name: verbs
src_addr: fi_sockaddr_in://10.12.6.80:0

prov_name: verbs;ofi_rxm
src_addr: fi_sockaddr_in://127.0.0.1:0

For FI_RMA I only get back verbs;ofi_rxm (on loopback).

prov_name: verbs;ofi_rxm
src_addr: fi_sockaddr_in://127.0.0.1:0

**Comment by bcernohous** [comment]
A simple mod to pingpong.c to test. (It doesn't actually run over RMA but I'm only interested in the fi_getinfo).

Output from 

$ fi_pingpong -p verbs -e rdm  -m msg  -S all -v  2>&1 | tee server.msg.out

$ grep -E "prov_name: verbs|src_addr:" server.*.out
server.msg.out:    src_addr: (null)
server.msg.out:        prov_name: verbs
server.msg.out:    src_addr: fi_sockaddr_in://10.12.6.80:0
server.msg.out:        prov_name: verbs

and (-r)

$ fi_pingpong -p verbs -e rdm  -m msg  -S all -v -r 2>&1 | tee server.rma.out

server.rma.out:    src_addr: (null)
server.rma.out:        prov_name: verbs
server.rma.out:    src_addr: fi_sockaddr_in://127.0.0.1:0
server.rma.out:        prov_name: verbs;ofi_rxm

diff --git a/util/pingpong.c b/util/pingpong.c
index e45df08..40d45b3 100644
--- a/util/pingpong.c
+++ b/util/pingpong.c
@@ -1436,6 +1436,11 @@ int pp_getinfo(struct ct_pingpong *ct, struct fi_info *hints,
                PP_PRINTERR("fi_getinfo", ret);
                return ret;
        }
+       PP_DEBUG("hint:fi_getinfo %s\n",
+                fi_tostr(hints, FI_TYPE_INFO));
+       PP_DEBUG("result:fi_getinfo %s\n",
+                fi_tostr(*info, FI_TYPE_INFO));
+
        return 0;
 }
 
@@ -1920,6 +1925,8 @@ void pp_pingpong_usage(char *name, char *desc)
        fprintf(stderr, " %-20s %s\n", "-m <transmit mode>",
                "transmit mode type: msg|tagged (msg)");
 
+       fprintf(stderr, " %-20s %s\n", "-r", "use FI_RMA instead of FI_MSG");
+
        fprintf(stderr, " %-20s %s\n", "-h", "display this help output");
        fprintf(stderr, " %-20s %s\n", "-v", "enable debugging output");
 }
@@ -1991,6 +1998,11 @@ void pp_parse_opts(struct ct_pingpong *ct, int op, char *optarg)
                ct->opts.dst_port = parse_ulong(optarg, UINT16_MAX);
                break;
 
+       case 'r':
+               ct->hints->caps &= ~FI_MSG;
+               ct->hints->caps |= FI_RMA;
+               break;
+              
        case 'm':
                if (strncasecmp("msg", optarg, 4)) {
                        ct->hints->caps &= ~FI_MSG;
@@ -2193,7 +2205,7 @@ int main(int argc, char **argv)
 
        ofi_osd_init();
 
-       while ((op = getopt(argc, argv, "hvd:p:e:I:S:B:P:cm:")) != -1) {
+       while ((op = getopt(argc, argv, "hvd:p:e:I:S:B:P:cm:r")) != -1) {
                switch (op) {
                default:
                        pp_parse_opts(&ct, op, optarg);

**Comment by gladkovdmitry17** [comment]
Thanks for reporting this!
I've checked the #3200 PR in my environment and it helps me to get all addresses except loopback for `verbs;ofi-rxm`.

Could you check too, please? Thanks in advance!

Btw, you may use `./fi_info -t FI_EP_RDM --verbose`. The `fi_info` utility prints all fields of the fi_info for each providers in "verbose" mode.

**Comment by bcernohous** [comment]
I see verbs on good addresses as well as verbs;ofi_rxm on loopback using fi_info, but not on the modified fi_pingpong.  I only get verbs;ofi_rxm on loopback in fi_pingpong.  

Are the hints being used forcing ofi_rxm?   Do we only support ofi_rxm on loopback?

$ bin/fi_info -t FI_EP_RDM --verbose | grep  -E "prov_name: verbs|src_addr:"
    src_addr: fi_sockaddr_in://10.12.6.80:0
        prov_name: verbs
    src_addr: fi_sockaddr_in://127.0.0.1:0
        prov_name: verbs;ofi_rxm
    src_addr: fi_sockaddr_in://127.0.0.1:0
        prov_name: verbs;ofi_rxm

$ bin/fi_pingpong -p verbs -e rdm  -m msg  -S all -v -r 2>&1 | tee server.rma.out

$ grep -E "prov_name: verbs|src_addr:" server.rma.out 
    src_addr: fi_sockaddr_in://127.0.0.1:0
        prov_name: verbs;ofi_rxm

$ git log --oneline | head
715b305 Merge pull request #3200 from a-ilango/rxm3
489f3c0 Merge pull request #3198 from pmmccorm/master
3e1594f Update version to 1.6.0a
1fa2995 prov/rxm: Fix an issue where fi_addr was not stored with connection info
ca4e982 v1.5.0
e096c13 NEWS.md: Add comments for bgq provider
1337327 Updated nroff-generated man pages
068fec0 Merge pull request #3201 from shefty/master
29aafe3 NEWS.md: Update news for netdir, mlx, sockets providers
05dd5d4 Do not cache a copy of struct fi_info internal to the verbs provider.

**Comment by a-ilango** [comment]
Hi, I see the issue on my end. There was a commit added to ofi_rxm provider that made it choose loopback if no address was given. This was to allow listening on loopback address if the user doesn't give any address info. I do have a patch that reverts this behavior since it wasn't playing well with MPICH.

Btw, if you give src or dest addr it shouldn't show loopback:
```
[arun@f017 ]$ fi_info -n f017-ib0 -t FI_EP_RDM --verbose | grep -E "prov_name: |src_addr:"
    src_addr: fi_sockaddr_in://192.168.133.17:0
        prov_name: verbs
    src_addr: fi_sockaddr_in://192.168.133.17:0
        prov_name: verbs;ofi_rxm
    src_addr: fi_sockaddr_in://192.168.133.17:0
        prov_name: verbs;ofi_rxm
    src_addr: fi_sockaddr_in://192.168.133.17:0
        prov_name: sockets
[arun@f017 ]$
```

Can you try this?

**Comment by bcernohous** [comment]
Yes, thanks.

I modified fi_pingpong to use a node name of 10.12.6.82 and that changed:

    src_addr: fi_sockaddr_in://127.0.0.1:0
        name: mlx5_0
        name: IB-0x10fe
        prov_name: verbs;ofi_rxm

into:

    src_addr: fi_sockaddr_in://10.12.6.82:0
        name: mlx5_0
        name: IB-0x10fe
        prov_name: verbs;ofi_rxm

**Comment by gladkovdmitry17** [comment]
@bcernohous Sorry, Have you found the answer to the question here? Shall we provide more detail? Or do you have any other questions related to this area?
Please, update us when we can close this issue. Thanks.

**Comment by bcernohous** [comment]
I thought the answer was you found a problem and had a patch?  I haven't tried to verify any patch.  

The work-around worked for me and I'm not waiting on anything more at this time.

**Comment by a-ilango** [comment]
This is fixed by a commit in #3260. I forgot this issue's number in the PR. I'll close this issue now.
